File: implementations/cpp/provekit/verifier/enumerate_callsites.cpp
// SPDX-License-Identifier: Apache-2.0

#include "enumerate_callsites.hpp"

namespace provekit::verifier {

namespace {

void walk_term(const Json& t,
               const std::string& property_name,
               const std::string& property_cid,
               const MementoPool& pool,
               std::vector<CallSite>& out);

void walk_formula(const Json& f,
                  const std::string& property_name,
                  const std::string& property_cid,
                  const MementoPool& pool,
                  std::vector<CallSite>& out) {
    if (!f.is_object()) return;
    const std::string kind = f.value("kind", "");
    if (kind == "atomic") {
        if (f.contains("args") && f["args"].is_array()) {
            for (const auto& a : f["args"]) walk_term(a, property_name, property_cid, pool, out);
        }
        return;
    }
    if (kind == "and" || kind == "or" || kind == "not" || kind == "implies") {
        if (f.contains("operands") && f["operands"].is_array()) {
            for (const auto& op : f["operands"]) walk_formula(op, property_name, property_cid, pool, out);
        }
        return;
    }
    if (kind == "forall" || kind == "exists") {
        if (f.contains("body")) walk_formula(f["body"], property_name, property_cid, pool, out);
        return;
    }
}

void walk_term(const Json& t,
               const std::string& property_name,
               const std::string& property_cid,
               const MementoPool& pool,
               std::vector<CallSite>& out) {
    if (!t.is_object()) return;
    if (t.value("kind", "") != "ctor") return;
    const std::string name = t.value("name", "");
    auto it = pool.bridges_by_symbol.find(name);
    if (it != pool.bridges_by_symbol.end()) {
        const auto& benv = it->second;
        const auto& bbody = benv["evidence"]["body"];
        CallSite cs;
        cs.bridge_ir_name = name;
        cs.bridge_target_cid = bbody.value("targetContractCid", "");
        cs.bridge_source_layer = bbody.value("sourceLayer", "");
        cs.bridge_target_layer = bbody.value("targetLayer", "");
        cs.property_name = property_name;
        cs.property_cid = property_cid;
        if (t.contains("args") && t["args"].is_array() && !t["args"].empty()) {
            cs.arg_term = t["args"][0];
        }
        out.push_back(std::move(cs));
    }
    if (t.contains("args") && t["args"].is_array()) {
        for (const auto& a : t["args"]) walk_term(a, property_name, property_cid, pool, out);
    }
}

}  // namespace

std::vector<CallSite> EnumerateCallsitesStage::Run(const MementoPool& pool) {
    std::vector<CallSite> out;
    for (const auto& [cid, env] : pool.mementos) {
        if (!env.contains("evidence") || !env["evidence"].is_object()) continue;
        const auto& ev = env["evidence"];
        if (ev.value("kind", "") != "contract") continue;
        if (!ev.contains("body") || !ev["body"].is_object()) continue;
        const auto& body = ev["body"];
        std::string property_name = body.value("contractName", "");
        if (property_name.empty()) {
            property_name = cid.substr(0, 12) + "...";
        }
        // Walk pre/post/inv (whichever are present). Each can independently
        // contain ctor invocations of bridge-source symbols (call sites).
        if (body.contains("pre") && body["pre"].is_object()) {
            walk_formula(body["pre"], property_name, cid, pool, out);
        }
        if (body.contains("post") && body["post"].is_object()) {
            walk_formula(body["post"], property_name, cid, pool, out);
        }
        if (body.contains("inv") && body["inv"].is_object()) {
            walk_formula(body["inv"], property_name, cid, pool, out);
        }
    }
    return out;
}

}  // namespace provekit::verifier

File: implementations/go/provekit-ir-symbolic/verifier/enumerate_callsites.go
package verifier

// EnumerateCallsitesStage walks every property memento in the pool,
// finds Ctor terms whose name matches a bridge envelope's sourceSymbol,
// and emits CallSite records — one per bridge call discovered inside
// a property memento's IR formula.
type EnumerateCallsitesStage struct{}

// Run produces CallSite records.
func (s *EnumerateCallsitesStage) Run(pool *MementoPool) []CallSite {
	var out []CallSite
	for cid, env := range pool.Mementos {
		ev, ok := env["evidence"].(map[string]interface{})
		if !ok || ev["kind"] != "property" {
			continue
		}
		body, _ := ev["body"].(map[string]interface{})
		if body == nil {
			continue
		}
		formula, _ := body["irFormula"].(map[string]interface{})
		if formula == nil {
			continue
		}
		var propertyName string
		if scope, ok := body["scope"].(map[string]interface{}); ok {
			if name, ok := scope["name"].(string); ok {
				propertyName = name
			}
		}
		if propertyName == "" {
			propertyName = cid[:12] + "…"
		}
		s.walkFormula(formula, propertyName, cid, pool.BridgesBySymbol, &out)
	}
	return out
}

func (s *EnumerateCallsitesStage) walkFormula(
	f map[string]interface{},
	propertyName, propertyCID string,
	bridges map[string]map[string]interface{},
	out *[]CallSite,
) {
	switch f["kind"] {
	case "atomic":
		if args, ok := f["args"].([]interface{}); ok {
			for _, a := range args {
				if at, ok := a.(map[string]interface{}); ok {
					s.walkTerm(at, propertyName, propertyCID, bridges, out)
				}
			}
		}
	case "and":
		if conjuncts, ok := f["conjuncts"].([]interface{}); ok {
			for _, c := range conjuncts {
				if cm, ok := c.(map[string]interface{}); ok {
					s.walkFormula(cm, propertyName, propertyCID, bridges, out)
				}
			}
		}
	case "or":
		if disjuncts, ok := f["disjuncts"].([]interface{}); ok {
			for _, d := range disjuncts {
				if dm, ok := d.(map[string]interface{}); ok {
					s.walkFormula(dm, propertyName, propertyCID, bridges, out)
				}
			}
		}
	case "not":
		if body, ok := f["body"].(map[string]interface{}); ok {
			s.walkFormula(body, propertyName, propertyCID, bridges, out)
		}
	case "implies":
		if a, ok := f["antecedent"].(map[string]interface{}); ok {
			s.walkFormula(a, propertyName, propertyCID, bridges, out)
		}
		if c, ok := f["consequent"].(map[string]interface{}); ok {
			s.walkFormula(c, propertyName, propertyCID, bridges, out)
		}
	case "forall", "exists":
		if pred, ok := f["predicate"].(map[string]interface{}); ok {
			if body, ok := pred["body"].(map[string]interface{}); ok {
				s.walkFormula(body, propertyName, propertyCID, bridges, out)
			}
		}
	}
}

func (s *EnumerateCallsitesStage) walkTerm(
	t map[string]interface{},
	propertyName, propertyCID string,
	bridges map[string]map[string]interface{},
	out *[]CallSite,
) {
	if t["kind"] != "ctor" {
		return
	}
	name, _ := t["name"].(string)
	bridgeEnv, ok := bridges[name]
	if ok {
		ev := bridgeEnv["evidence"].(map[string]interface{})
		body := ev["body"].(map[string]interface{})
		args, _ := t["args"].([]interface{})
		var firstArg interface{}
		if len(args) > 0 {
			firstArg = args[0]
		}
		*out = append(*out, CallSite{
			BridgeIRName:      name,
			BridgeTargetCID:   asString(body["targetContractCid"]),
			BridgeSourceLayer: asString(body["sourceLayer"]),
			BridgeTargetLayer: asString(body["targetLayer"]),
			PropertyName:      propertyName,
			PropertyCID:       propertyCID,
			ArgTerm:           firstArg,
		})
	}
	// Recurse into ctor args.
	if args, ok := t["args"].([]interface{}); ok {
		for _, a := range args {
			if am, ok := a.(map[string]interface{}); ok {
				s.walkTerm(am, propertyName, propertyCID, bridges, out)
			}
		}
	}
}

func asString(v interface{}) string {
	if s, ok := v.(string); ok {
		return s
	}
	return ""
}
